### ISSUE TITLE
libcurl 7.88.0- getaddrinfo() thread failed to start

### ISSUE THREAD

**Issue by gunamoi1** [curl/curl#10561]
I use libcurl, compiled as a static library, in many of my programs.
Code written in C++, for Windows, using Visual Studio.
libcurl static library is compiled to call DLL Windows SSPI, and DLL WinIDN.
I have used all previous versions from 7.50.3 to 7.86.0, and all works great.

Recently I saw 7.88.0 is available, so I downloaded and compiled that one, using the same settings as all my previous builds.
Now all of my programs fail, with a message:
"getaddrinfo() thread failed to start".

So, I have gone back to using the previous version 7.86.0, and all my programs work again.

What might be causing this bug in 7.88.0?
Is there some extra curl setup required in this new version which I will have to change in my programs to make them work with 7.88.0?

**Comment by jay** [comment]
Ref: https://github.com/curl/curl/issues/6495
Ref: https://github.com/curl/curl/issues/10144

Possible resource issue or MITM issue. Allowing MITM was supposed to be fixed in 7.88.0 unless that broke something for you. Does it work if you use the curl tool (curl.exe)? Does it happen 100% of the time? If you go back to 7.86 does it work? If you git checkout curl-7_88_0 and then revert 7fa449ca does it work? Do you have an anti-virus client that could be interfering?

You could further narrow it down by [bisecting](https://github.com/curl/curl/wiki/how-to-git-bisect). The 'Build but no install' step is specific to autotools builds so ignore it in this case. Since you are using a VS project configuration (eg 'DLL Release - DLL Windows SSPI - DLL WinIDN') your 'Build but no install' step would be run `generate.bat <vcver>` (eg `generate vc14.30`) from the projects directory then build the configuration in Visual Studio.

**Comment by u20221022** [comment]
> Does it work if you use the curl tool (curl.exe)? Does it happen 100% of the time? … Do you have an anti-virus client that could be interfering?

I got the same message in command line a few hours ago when I was preparing examples for #10558.
Only once.  MS Windows Defender only.  MS Windows 10.

**Comment by jay** [comment]
I have been able to reproduce approx every twenty minutes but only in certain circumstances. No debugging or printfs. OutputDebugString ok though. I run the same command repeatedly in a batch file, with apologies to the apache folks, until it fails:

~~~batch
FOR /L %%G IN (1,1,%MAX%) DO (
  curld -sS -f -H "Connection: close" https://httpd.apache.org/ -o NUL --write-out "%%{time_total} "
  if !ERRORLEVEL! NEQ 0 exit /b 1
)
~~~

It looks like it is related to 7fa449ca.

https://github.com/curl/curl/blob/41dfb7f516c8a9659782df695528c9d32b5653d6/lib/socketpair.c#L118-L124

sread returns SOCKET_ERROR and WSAGetLastError returns WSAEWOULDBLOCK. The socket is non-blocking so we will have to take into account that sread/recv (and to a lesser extent swrite/send) may return this. @bagder

**Comment by u20221022** [comment]
MITM at fake addresses curl1.se/curl2.se/…?
And why error №6 («Could not resolve host.»), not №2 («Failed to initialize.»)?

**Comment by bagder** [comment]
@gunamoi1 does #10562 fix the issue for you?

**Comment by renodr** [comment]
Hi folks,

I was able to determine that this fixes the problem and posted my procedure/results at [https://curl.se/mail/lib-2023-02/0064.html](url)

**Comment by jay** [comment]
Thanks for testing. It depends a lot on system resources and how many sockets are active at the same time. I've made a branch with all the debug statements that I added and a better repeat command using `http://curl.se`, since it is just a redirect that is not followed and therefore the transfer completes much faster making it possible to repro in less than 5 minutes here.

[jay:curl:win32_debug_socketpair_init](https://github.com/curl/curl/compare/master...jay:curl:win32_debug_socketpair_init?expand=1)

[Capture1](https://user-images.githubusercontent.com/965580/219975858-8dbc4da3-c523-4e07-bf04-30fbf44fbdc4.PNG)
[Capture2](https://user-images.githubusercontent.com/965580/219975869-479d2e7d-82d8-4b66-aa62-db0d75bef59d.PNG)
[Capture3](https://user-images.githubusercontent.com/965580/219975930-2b7a6f57-68fd-4d4d-a6a1-3768dc5f976d.PNG)

I have yet to investigate #10562

**Comment by bagder** [comment]
Thanks a lot @renodr!

**Comment by renodr** [comment]
Just for some followup! 😄 

@jay I pulled a copy of your batch file down and edited it to go to linuxfromscratch.org (since I have access to that server) and then did a fresh build of cURL in a separate directory so that I could have a build available without #10562 applied to it.

On the unpatched build, I got the following output:

![image](https://user-images.githubusercontent.com/11901488/219978167-7f2573f4-9fae-4423-b635-73ee2962cb61.png)

On the patched build, I get:

![image](https://user-images.githubusercontent.com/11901488/219981923-e2701ba8-2238-48a7-b4e1-6a0fcfa51e1a.png)

It ran for about an hour and 20 minutes without issues, the only reason why it stopped was because I had a network outage

I think this should be good to go! 😄

**Comment by vszakats** [comment]
Test curl-for-win binaries:
https://ci.appveyor.com/project/curlorg/curl-for-win/builds/46278505/artifacts

(disregard the version number in the filename, this is latest master + patch #10562)

**Comment by gunamoi1** [comment]
@jay 
Yes, problem happens with my apps 100% of the time if compiled with 7.88.0.
If recompiled with 7.86.0, they all work perfectly.

I have only tested on one machine, Windows 10.
Even when anti-virus and firewall are disabled, v7.86 works, and v7.88.0 fails.

I downloaded a daily release, (curl-7.88.1-20230219.zip)
but that still has the same problem as 7.88.0.

@bagder
I am not a github power user.
I tried downloading and compiling #10562, (curl-bagder-socketpair-block.zip)
but get a bunch of fatal errors about missing files when I try to compile LIB Release – DLL Windows SSPI – DLL WinIDN
missing files:
..\lib\idn_win32.c
..\lib\vquic\msh3.c
..\lib\vquic\ngtcp2.c
..\lib\vquic\quiche.c
I don't get these errors when compiling the master 7.86.0, or 7.88.0, so maybe I did something wrong downloading #10562.

Anyway, based on the tests by others, it looks like your solution will probably fix my issue.
But I'll probably have to wait until it is included in a full master release to confirm myself.
regards,

**Comment by vszakats** [comment]
@gunamoi1: You can test the patch with these binaries: https://ci.appveyor.com/api/buildjobs/tcl16t5jhb3neqdj/artifacts/all-mingw-7.88.0.zip

**Comment by gunamoi1** [comment]
@vszakats 
Thanks, but I'm compiling into my programs as static library (LIB) not as DLL.

@bagder 
I managed to get something that would compile, by starting with the contents of curl-7.88.0.zip master,
and then copying just the source files from lib and src folders of #10562 (from curl-bagder-socketpair-block.zip).
Probably just needed socketpair.c, but I copied them all anyway.

Once compiled into a static LIB and the compiled into my programs, all works good and the original problem goes away.
So, many thanks, you've definitely fixed this issue with #10562 as far as I'm concerned.
